File: kattis/insert/insert.cpp
#include <iostream>

using namespace std;

typedef long long int ll;

// 3 4 3 5 4 1 2
struct bst_node {
    int value;
    ll left_size, right_size;
    bst_node *left, *right;

    bst_node(int _value) : value(_value) {
        left_size = right_size = 0;
        left = NULL;
        right = NULL;
    }
};

bst_node *insert(bst_node *root, int v) {
    if (!root) return new bst_node(v);

    if (v >= root->value) {  // to the right
        root->right_size++;
        root->right = insert(root->right, v);
    } else {  // to the left
        root->left_size++;
        root->left = insert(root->left, v);
    }

    return root;
}

// Number of ways to choose k objects from a set of n objects, ignoring order.
ll binomial(ll n, ll k) {
    if (k == 0 || k == n) return 1;
    return binomial(n - 1, k - 1) + binomial(n - 1, k);
}

ll count_binomial(bst_node *root) {
    if (!root) return 1;

    return binomial(root->left_size + root->right_size, root->right_size) *
           count_binomial(root->left) * count_binomial(root->right);
}

int main() {
    int n, aux;

    while (cin >> n && n != 0) {
        bst_node *tree;
        for (int i = 0; i < n; i++) {
            cin >> aux;
            tree = insert(tree, aux);
        }

        cout << count_binomial(tree) << endl;
        tree = NULL;
    }

    return 0;
}

File: kattis/insert/insert.rb
class Node
  attr_accessor :value, :left, :right, :left_size, :right_size

  def initialize(v)
    @value = v
    @left_size = 0
    @right_size = 0
  end

  def insert(v)
    if value
      if v >= value
        self.right_size+=1
        insert_right(v)
      else
        self.left_size+=1
        insert_left(v)
      end
    else
      self.value = v
    end
  end

  def insert_left(v)
    if left
      left.insert(v)
    else
      self.left = Node.new(v)
    end
  end

  def insert_right(v)
    if right
      right.insert(v)
    else
      self.right = Node.new(v)
    end
  end
end

@dynbin = Hash.new { |hash, key| hash[key] = {} }

def binomial(n, k)
  return 1 if k==0 || k==n

  @dynbin[n-1][k-1] ||= binomial(n-1, k-1)
  @dynbin[n-1][k] ||= binomial(n-1, k)

  @dynbin[n-1][k-1] + @dynbin[n-1][k]
end

def count_binomial(root)
  return 1 if !root

  binomial(root.left_size + root.right_size, root.right_size) * count_binomial(root.left) * count_binomial(root.right)
end

n = 0
$stdin.each_line do |line|
  if n==0
    n = line.to_i
  else
    root = Node.new(nil)

    line.split(' ').each do |value|
      root.insert(value.to_i)
    end

    puts count_binomial(root)

    n=0
  end
end
